### PR TITLE
Add retry handling for AWS providers

### PR DESCRIPTION
**Pull request by thomaschangsf**
Pull requests adds retry logic on AWS IO operators.

Risk: low

### CODE REVIEW

**Review by damianxu88 — APPROVED** [review]
LGTM

**Comment by schowsf** [comment]
I want to get your opinion on moving Retry 
from orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/util/Retry.scala
to orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala

Some other future orchard providers can also use it.  I see from the build.sbt, orchardProviderAWS dependsOn(orchardCore) already.

**Comment by damianxu88** [comment]
@thomaschangsf we could look at using https://github.com/salesforce/stubborn/ directly also which has exponential backoff built-in since majority of the calls are for AWS APIs in Orchard.

**Comment by thomaschangsf** [comment]
> 

I had the same thought early on, but decided to keep the scope of PR changes smaller.  But, I agree with your comment and will move the Retry to the core project makes sense for future IO providers.

**Comment by realstraw** [comment]
How does using suborn make the scope change larger?

**Comment by thomaschangsf** [comment]
> How does using suborn make the scope larger? Isn't it would make it smaller given it's a library that we maintain and already have all the necessary policy built-in?

My comment about keeping the scope smaller was in response to Simon's comment to move the util.Retry from orchardProviderAWS project to orchardCore project since this PR is for adding retry logic to AWS clients.

I agree this PR should incorporate the stubborn git repository, and this afternoon's update incorporates it.  I would, however, appreciate your thoughts on how best to integrate the stubborn library.  The two options I see are:
(1) Extend all AWS clients classes with the Retryable and Policy traits, ie hyperion repo
(2) Wrap the stubborn library inside a Retry helper object, ie this afternoon's commit

Both has its pros and cons; however approach 1 requires refactoring to
(A) Client.scala's factory methods for ec2, emr, and ssm to multiple new classes
(B) The classes inside the resource and activity packages to the new classes 

This would increase the scope of changes in this PR.  Curious if you see a simpler more elegant alternative.

**Comment by realstraw** [comment]
> My comment about keeping the scope smaller was in response to Simon's comment to move the util.Retry from orchardProviderAWS project to orchardCore project since this PR is for adding retry logic to AWS clients.

Cool, got it.

> I agree this PR should incorporate the stubborn git repository, and this afternoon's update incorporates it. I would, however, appreciate your thoughts on how best to integrate the stubborn library. The two options I see are:
> (1) Extend all AWS clients classes with the Retryable and Policy traits, ie hyperion repo
> (2) Wrap the stubborn library inside a Retry helper object, ie this afternoon's commit

I don't mind either and subborn supports both patterns IIRC. It's up to you, I'm not too sure if option 1) would result in that much refactoring, but I'd say option 2) is pretty clean implementation too.

File: orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+object Retry extends RetryDefaults {
+  @annotation.tailrec
+  def apply[T](numTries: Int = defaultMaxRetries, policy: Policy = defaultPolicy)(
+    fn: => T
+  ): Try[T] = {
+    Try { fn } match {
+      case Failure(_) if numTries > 1 =>
+        val delay = policy.retryDelay(defaultMaxRetries - numTries)
+        Thread.sleep(delay)
```
Comment (realstraw) [inline orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala:20]: you don't have to reimplement those. stubborn already provided them

Comment (thomaschangsf) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala:20]: Good idea.  I will use stubborn library's RetryDefaults.

File: build.sbt [hunk 2]
```diff
@@ -1,8 +1,11 @@
+
+
```
Comment (realstraw) [inline build.sbt:2]: unnecessary new lines

File: orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.krux.stubborn.RetryDefaults
+import com.krux.stubborn.policy.Policy
+
+import scala.util.{Failure, Try}
+
+object Retry extends RetryDefaults {
+
+  @annotation.tailrec
+  def apply[T](numTries: Int = defaultMaxRetry, policy: Policy = defaultPolicy)(
```
Comment (realstraw) [inline orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala:11]: you can just use this one, no? https://github.com/salesforce/stubborn/blob/master/src/main/scala/com/krux/stubborn/Retryable.scala

Comment (thomaschangsf) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala:11]: Yes, you are right. I will use Retryable.retry.

File: orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala [hunk 4]
```diff
@@ -21,51 +24,55 @@ import com.salesforce.mce.orchard.system.util.InvalidJsonException
...
-          .jobFlowId(clusterId)
-          .steps(
-            steps.map { step =>
-              StepConfig
-                .builder()
-                .name(name)
-                .actionOnFailure(ActionOnFailure.CONTINUE)
-                .hadoopJarStep(
-                  HadoopJarStepConfig
+  override def create(): Either[Throwable, JsValue] = catching(classOf[SdkException]) either
```
Comment (realstraw) [inline orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala:27]: I think we should return everything as Either, not just SdkException.

File: orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala (hunk 3 above)
Comment (realstraw) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala]: there's nothing stops you from writing a function that uses `Retryable` to return `Try`/`Either`.

File: orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala (hunk 4 above)
Comment (thomaschangsf) [reply orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala:27]: Sure. Do you think RunTimeException would be sufficient?

File: orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala (hunk 3 above)
Comment (thomaschangsf) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala]: That is correct.

Comment (thomaschangsf) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala]: Please  review this stubborn PR, which adds a method to wrap Retryable.retry() in a Try:
https://github.com/salesforce/stubborn/pull/4

Comment (realstraw) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala]: you don't have to change stubborn. have a method that wraps try here.

Comment (thomaschangsf) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala]: I was thinking this retry in a Try wrapper idiom would be common across projects. 

Sure, I can add a method in this project to wrap a Try around retry().

File: orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala (hunk 4 above)
Comment (realstraw) [reply orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala:27]: yeah, basically, what's wrong with `Try(...).toEither`?

File: orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala (hunk 3 above)
Comment (thomaschangsf) [reply orchard-core/src/main/scala/com/salesforce/mce/orchard/util/Retry.scala]: I crated a method to wrap Retryable.retry in a Try in the latest commit.  Please let me know what you think.

File: orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala (hunk 4 above)
Comment (thomaschangsf) [reply orchard-provider-aws/src/main/scala/com/salesforce/mce/orchard/io/aws/activity/EmrActivity.scala]: I agree Try(...).toEither is better than the catching idiom.  Latest commit refactor to this idiom.

**Review by realstraw — APPROVED** [review]

**Pull request merged**